app/services/UnitLinksEditService.scala

```diff
@@ -15,24 +15,45 @@ import scala.concurrent.ExecutionContext.Implicits.global
 
 class UnitLinksEditService @Inject() (repository: AdminDataUnitLinksEditRepository) extends EditService with LazyLogging {
 
+  private case class AddAndDeleteOperationsParams(createUnitKey: UnitKey, deleteUnitKey: UnitKey, vatIdAndType: IdAndType)
+
   override def editVatParentUnitLink(period: Period, vatref: VatRef, editParentLink: EditParentLink): Future[EditParentLinkStatus] = {
     val parent = editParentLink.parent
-
     val updateParentUnitKey = UnitKey(UnitId(vatref.value), ValueAddedTax, period)
     val createChildUnitKey = UnitKey(parent.to.id, LegalUnit, period)
     val deleteChildUnitKey = UnitKey(parent.from.id, LegalUnit, period)
-
     val vatIdAndType = (IdAndType.apply _).tupled(VatUnitRef.toIdTypePair(vatref))
+    val addAndDeleteOperationParams = AddAndDeleteOperationsParams(createChildUnitKey, deleteChildUnitKey, vatIdAndType)
 
     repository.updateVatParentUnitLink(updateParentUnitKey, parent.from, parent.to).flatMap {
-      case PatchSuccess => Future.sequence(Seq(
-        repository.createLeuChildUnitLink(createChildUnitKey, vatIdAndType),
-        repository.deleteLeuChildUnitLink(deleteChildUnitKey, vatIdAndType)
-      )).map(reducePatchStatuses)
+      case PatchSuccess => submitAddAndDeleteOperations(addAndDeleteOperationParams)
+      case PatchConflict => handlePossibleConflict(updateParentUnitKey, parent, addAndDeleteOperationParams)
       case failedStatus => Future.successful(failedStatus)
     } map patchStatusToEditStatus
   }
 
+  private def submitAddAndDeleteOperations(params: AddAndDeleteOperationsParams): Future[PatchStatus] =
+    Future.sequence(Seq(
+      repository.createLeuChildUnitLink(params.createUnitKey, params.vatIdAndType),
+      repository.deleteLeuChildUnitLink(params.deleteUnitKey, params.vatIdAndType)
+    )).map(reducePatchStatuses)
+
+  /**
+   * When the user submits an editVatParentUnitLink request, if the update operation succeeds but any of the
+   * subsequent create/delete operations fail, the request will need to be retried by the user, to maintain
+   * data consistency in HBase.
+   *
+   * If the user re-submits the request (with the same JSON), as the first update operation succeeded before, that
+   * same operation now returns a conflict as the from value has been updated. In this scenario, we submit the same
+   * request, just with the from/to value being the updated (to) value, thus the request will succeed and the
+   * subsequent add/delete operations can be submitted.
+   */
+  private def handlePossibleConflict(updateUnitKey: UnitKey, parent: Parent, addDeleteParams: AddAndDeleteOperationsParams): Future[PatchStatus] =
+    repository.updateVatParentUnitLink(updateUnitKey, parent.to, parent.to) flatMap {
+      case PatchSuccess => submitAddAndDeleteOperations(addDeleteParams)
+      case failedStatus => Future.successful(failedStatus)
+    }
+
```

Comment:
This works, and I am not proposing that we change it.
However, note that there are now **two** different places in this code where ```submitAddAndDeleteOperations``` is called - this fact is the reason why you need something like ```AddAndDeleteOperationParams``` so that you can ship parameters to somewhere else in the code to make the same call you actually make in one of the cases of your current function.

Conceptually however, we only ever want to call ```submitAddAndDeleteOperations``` once, and so ideally, there should only be one place in the code where it is called.  This greatly simplifies readability / debugging etc.

I would have treated the update operation as a "logical operation".  Only on a success result from this, would I have requested addAndDeleteOperations.  This would therefore only be used once in the code.  

The function presenting a "logical update operation" would have handled making the update request, and chaining the follow up dummy update only in the scenario where the first returned conflict.  You could then safely call this function and know that if it returned success, you should proceed to the addAndDeleteOperations step, otherwise you should abort with the returned error code.

Reply:
That definitely is a lot more intuitive, I have changed my code to reflect your comments.